Enabling Selective Documentation Builds.

WarpWing opened the pull request:
**Pull Request Description**

Aims to implement suggestions in #94 

Closes #94 

---

**Licensing Certification**

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request **I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)** for its contents.

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
Currently did single target builds, working on adding selective as a whole. Thoughts appreciated!

WarpWing committed:
Added comments back

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

braughtg commented:
This looks like a good approach to building everything or just one individual component or library, which I expect would be the majority of use cases.

One other possibility would be to use globbing to build a collection of the docs. I don't know that that is necessary at this point, but if you want to have a go at it that would be fine.  Otherwise, flip this from draft and let me know and I'll review it.

WarpWing commented:
> This looks like a good approach to building everything or just one individual component or library, which I expect would be the majority of use cases.
> 
> One other possibility would be to use globbing to build a collection of the docs. I don't know that that is necessary at this point, but if you want to have a go at it that would be fine. Otherwise, flip this from draft and let me know and I'll review it. 

Thanks for the speedy response. Happy New Years! In addition, thoughts on doing selective components or just one single  or all of them? Haven't heard of "globbing" before, I assume its just stick it all together but I can google.

braughtg commented:
Happy New Year to you as well!

> thoughts on doing selective components or just one single or all of them?
I think doing all of the docs at once will be something people do when they setup the dev environment or re-sync with the upstream.  Then when doing work, the most common thing will be to generate the docs for a specific component or library on which they are working.  So, I would be happy with something that covers just those two use cases without being more general at this point.

> Haven't heard of "globbing" before, I assume its just stick it all together but I can google.
It expands the idea of wild cards to allow flexible specification of multiple files/directories at once.  It could allow for things like generating docs for all of the components or all of the libraries at once.

WarpWing commented:
`makeDocs.bash` should now accepts a command line argument to specify a component or library for documentation generation, or generates documentation for all components and libraries if no argument is provided.

WarpWing committed:
Update makeDocs.bash to use --target

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
To generate documentation for a specific component: ./makeDocs.bash --target componentName
    
    To generate documentation for multiple specific components: ./makeDocs.bash --target componentOne --target componentTwo
    
    To generate documentation for all components and libraries: ./makeDocs.bash (without any --target argument)
    
    
  I did multiple specific components for fun, wasn't necessary but who knows what edge cases this could solve.

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

braughtg reviewed (changes requested):
Running without the `--target` flag correctly generates all of the documentation.

Running with a `--target` flag for selecting a specific target (or targets) correctly generates the docs for the targets.

However before doing so, all of the other documentation is deleted.  So we end up with only the targeted documentation.  It would be preferable to leave all non-targeted docs in place, and just (re)generate the targeted docs.

For example, when I run: `makeDocs.bash --target NumericInput --target DateSelector
`

I end up with the following docs page:

<img width="885" alt="image" src="https://github.com/FarmData2/FarmData2/assets/16962482/372525ed-5f7b-45a3-a563-6785da52a818">

rather than a full docs page with the `NumericInput` and `DateSelector` docs updated.

My suggestion would be that when `--target` flags are included you don't delete all of the docs at the start (as the original script did).  Instead, leave the existing docs in place and just rebuild the full `FarmData2.md` file as you iterate over the full set of components and libraries.  Then when a targeted component or library is encountered, delete the targeted docs (if they exist) and (re)build them.  You may have to be careful here for the case where the targeted library or component is new (i.e. not docs exist for it yet).

The end result of running the script should be a full `FarmData2.md` document regardless of whether individual components or libraries are targeted.

WarpWing committed:
Added Component and Library Doc Functions

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
> Running without the `--target` flag correctly generates all of the documentation.
> 
> Running with a `--target` flag for selecting a specific target (or targets) correctly generates the docs for the targets.
> 
> However before doing so, all of the other documentation is deleted. So we end up with only the targeted documentation. It would be preferable to leave all non-targeted docs in place, and just (re)generate the targeted docs.
> 
> For example, when I run: `makeDocs.bash --target NumericInput --target DateSelector `
> 
> I end up with the following docs page:
> <img alt="image" width="885" src="https://private-user-images.githubusercontent.com/16962482/296258954-372525ed-5f7b-45a3-a563-6785da52a818.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0Mjk5MjUsIm5iZiI6MTcwNTQyOTYyNSwicGF0aCI6Ii8xNjk2MjQ4Mi8yOTYyNTg5NTQtMzcyNTI1ZWQtNWY3Yi00NWEzLWE1NjMtNjc4NWRhNTJhODE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE4MjcwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMTQxNTI2ZDI3NzZjMjg1OGVkOTUzM2NmMWYzMDJhMWViZjBhOGEzMWI2OGEzMTg2YjI2OGZjNzQ2NDBlYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fViWpn3mwyMqc3UWmkw5jjdD3rE-AJGwbCgKKVQHMis">
> 
> rather than a full docs page with the `NumericInput` and `DateSelector` docs updated.
> 
> My suggestion would be that when `--target` flags are included you don't delete all of the docs at the start (as the original script did). Instead, leave the existing docs in place and just rebuild the full `FarmData2.md` file as you iterate over the full set of components and libraries. Then when a targeted component or library is encountered, delete the targeted docs (if they exist) and (re)build them. You may have to be careful here for the case where the targeted library or component is new (i.e. not docs exist for it yet).
> 
> The end result of running the script should be a full `FarmData2.md` document regardless of whether individual components or libraries are targeted.

Ah, I see. I've made a few changes. I've deleted the beginning where it deleted all existing docs at the beginning and checked `FarmData2.md`'s existence so the script doesn't overwrite it. I rewrote the component and library document generation into modular functions instead which should be nice and readable.

braughtg reviewed (approved):
I like the idea of factoring out the doc creation pieces into functions - nicely done!

Overall it is getting closer but still seems to have some issues related to the links in the index document `FarmData2.md`.  It looks like the code for adding the links for the docs was removed and not re-added anywhere else.

My suggestion would be to completely delete and recreate the `FarmData2.md` file on each run.  Then as you are iterating over the directories for the components (or the libraries) decide if a link should be added for the current component.  A link should be added if either a document already exists for the component or the current run is generating the document for that component. 

The end result of every run of `makeDocs.bash` should be that there is a full `FarmData2.md` file with links to the docs for all of the components and all of the library for which docs have been generated.

WarpWing committed:
Merge branch 'development' into patch-3

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Added back clarifying comments

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Merge branch 'development' into patch-3

WarpWing commented:
> I like the idea of factoring out the doc creation pieces into functions - nicely done!
> 
> Overall it is getting closer but still seems to have some issues related to the links in the index document `FarmData2.md`. It looks like the code for adding the links for the docs was removed and not re-added anywhere else.
> 
> My suggestion would be to completely delete and recreate the `FarmData2.md` file on each run. Then as you are iterating over the directories for the components (or the libraries) decide if a link should be added for the current component. A link should be added if either a document already exists for the component or the current run is generating the document for that component.
> 
> The end result of every run of `makeDocs.bash` should be that there is a full `FarmData2.md` file with links to the docs for all of the components and all of the library for which docs have been generated.

Alright, I've implemented your suggestion to delete (technically overwrite) and regenerate the `FarmData2.md` file. I also isolated the building of documentation by vue-docgen or jsdoc2md on type in the new `update_docs()` function instead of `update_component_docs()` and `update_library_docs()` for modularity. This should be clear, waiting for final review!

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,72 @@
-  echo "- [$COMP_NAME]($COMP_MD_LINK) - $DESC_TEXT" >> "$INDEX_PATH"
-  echo "      Added."
 
-  echo "      Adding back links from $COMP_MD_FILE to $INDEX_FILE..."
```
The links from the doc pages back to the main page should still be included.  You'll want to look at this section and add those back in.

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,72 @@
-  echo "      Created."
-
-  echo "      Adding link for $COMP_MD_FILE to $INDEX_FILE..."
-  DESC_TEXT=$(grep -m 1 -A 1 "/\*\*" "$COMP_VUE_PATH" | tail -1 | cut -d' ' -f3-)
```
The links to each component or library should be followed by the first line of the description from their docs as was done here.

WarpWing replied:
Added back, I also reimplemented it to use head instead of grep. Unless I'm missing something, this should be much faster.

WarpWing replied:
```bash
# Extracts the first line of the description
local DESCRIPTION=$(head -n 1 "$REPO_ROOT_DIR/$DOC_PATH")

# Adds link with description to the index file
echo "- [$NAME]($DOC_PATH) - $DESCRIPTION" >> "$INDEX_PATH"

# Adds back link to the documentation file
echo "" >> "$REPO_ROOT_DIR/$DOC_PATH"
echo "[Back to Index](../$INDEX_FILE)" >> "$REPO_ROOT_DIR/$DOC_PATH"
```

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,72 @@
+        echo "    Generating docs for $NAME..."
+        # Use vue-docgen or jsdoc2md based on type
+        if [ "$TYPE" == "components" ]; then
+            npx vue-docgen "$PATH/$NAME.vue" -o "$REPO_ROOT_DIR/$DOC_PATH"
```
For some reason I'm getting the following error here (and similarly on line 43:

<img width="819" alt="image" src="https://github.com/FarmData2/FarmData2/assets/16962482/d71cafce-920b-45cc-ba76-92b8eebf5a5d">

This results in no doc (`.md`) files being generated.

But if I run `npx` from the command line it works fine.  Are you experiencing anything similar?

WarpWing replied:
I think I've figured it out why... the error with the .md files not being generated might be related to the working directory or environment when npx is being called. Since it works fine when you run npx from the command line, it suggests that the issue is with how the environment is set up in the script. I'm going to chain the command to first `cd` from the root dir like so `(cd "$REPO_ROOT_DIR" && npx vue-docgen "$PATH/$NAME.vue" -o "$DOC_PATH")` and the same thing for jsdoc2md

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,72 @@
+
+TARGETS=()
+
+# Parsing all command line arguments
```
It would be nice if the default behavior, with no arguments, were to generate all of the documentation.

WarpWing replied:
Created an adhoc conditional to do so.

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,72 @@
+        echo "      Docs generated for $NAME."
+
+        # Adds link to the index file
+        echo "- [$NAME]($DOC_PATH)" >> "$INDEX_PATH"
```
The links seem like they may be off... when visit the `FarmData2.md` page there are links for each component and library. However clicking on any of the links results in something like:

<img width="690" alt="image" src="https://github.com/FarmData2/FarmData2/assets/16962482/37a16c7c-424c-4ea2-9c97-0ee958e03192">

There seems to be an extra `docs` in there.

WarpWing replied:
Fixed Doc Path to be `local DOC_PATH="$TYPE/$NAME.md"`

braughtg reviewed (changes requested):
Overall this is improving and getting closer.  I've added some comments directly to the code pointing out some areas where there are things to be addressed.

What I am seeing though is that there is now no documentation being generated when I run `makeDocs.bash --target all`. 

Have you been able to run it and successfully generate all of the docs?

WarpWing commented:
> Overall this is improving and getting closer. I've added some comments directly to the code pointing out some areas where there are things to be addressed.
> 
> What I am seeing though is that there is now no documentation being generated when I run `makeDocs.bash --target all`.
> 
> Have you been able to run it and successfully generate all of the docs?

I think I've made some adhoc changes between my version and the git version. I have been able to generate with just `--target` which generates all of the documents. I'll look over the changes and respond to all of them. Thanks for the review!

braughtg commented:
> I have been able to generate with just `--target` which generates all of the documents. 

Ahh... I tried `makeDocs.bash --target=all`

Might be worth adding a `--help` flag as well that explains usage.  You can find some examples of that in the other scripts in `bin`.

> I'll look over the changes and respond to all of them. Thanks for the review!

Happy to have the help and do the reviews!  Thanks for the contributions!

WarpWing commented:
> I have been able to generate with just `--target` which generates all of the documents.
> 
> Ahh... I tried `makeDocs.bash --target=all`
> 
> Might be worth adding a `--help` flag as well that explains usage. You can find some examples of that in the other scripts in `bin`.
> 
> > I'll look over the changes and respond to all of them. Thanks for the review!
> 
> Happy to have the help and do the reviews! Thanks for the contributions!

Logically speaking, that functionality makes sense and it should be added. I can implement the `--help` function and make it use `--target=all` as well. I will look into npx and also why it's adding in a docs duplicate in the page link as well, that's strange.

Regarding npx, is it possible that it's not seeing the direct bin executable? Maybe if npx is installed per project, it doesn't pick it up. I'm not sure as to that. I'll fix everything else and have it reviewable in the afternoon.

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
Implemented all of the changes. I also did a bit of work to rework the backlinks and descriptions to this: 

```bash
# Extracts the description using grep from the .vue file
local DESCRIPTION=$(grep -m 1 -A 1 "/\*\*" "$VUE_PATH" | tail -1 | cut -d' ' -f3-)

# Adds link with description to the index file
echo "- [$NAME]($DOC_PATH) - $DESCRIPTION" >> "$INDEX_PATH"

# Adds back link to the documentation file
echo "" >> "$REPO_ROOT_DIR/$DOC_PATH"
echo "[Back to Index](../$INDEX_FILE)" >> "$REPO_ROOT_DIR/$DOC_PATH"
      ```

WarpWing commented:
If you would like me to make a `--help`` command, I very much can, but I would like to fix all bugs before moving on with enhancements.

WarpWing commented:
Apologies for not re-requesting a review, I was over at Georgetown doing a Hackathon with Boosung and some friends. I'm back and I think I'll start looking at a help command!

braughtg commented:
> Apologies for not re-requesting a review, I was over at Georgetown doing a Hackathon with Boosung and some friends. I'm back and I think I'll start looking at a help command!

Cool stuff!  Hope it was fun!

I'll. have a look at this in the next few days.  In the meantime, feel free to go on to the help feature if you have time and interest.

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,83 @@
-  echo "      Created."
-
-  echo "      Adding link for $COMP_MD_FILE to $INDEX_FILE..."
-  DESC_TEXT=$(grep -m 1 -A 1 "/\*\*" "$COMP_VUE_PATH" | tail -1 | cut -d' ' -f3-)
```
This command gets the first line of the JSDOC, not the first line of the file.  The first line of the JSDoc provides a description of the component or library and is what should appear in the docs.  So you'll want to use this command rather than just the `head` command you used above.

braughtg commented on bin/makeDocs.bash:
```diff
@@ -3,118 +3,83 @@
+        echo "      Docs generated for $NAME."
+
+        # Extracts the first line of the description
+        local DESCRIPTION=$(head -n 1 "$REPO_ROOT_DIR/$DOC_PATH")
```
See comment below about the command you need here.

WarpWing replied:
I see, my apologies, I thought I could just shortcut it, I'll make sure to reimplement and test

braughtg reviewed (changes requested):
I feel like this is getting close, but I am still seeing some issues.  I left a few comments in the code about specific things.  Also when I run the program I'm seeing the following errors:

<img width="826" alt="image" src="https://github.com/FarmData2/FarmData2/assets/16962482/fb10e3c5-e10c-40fd-b0bd-42a93cdbf92c">

However, both `npx` and `head` exist and run just fine.  

<img width="826" alt="image" src="https://github.com/FarmData2/FarmData2/assets/16962482/9f2bc619-dd23-400a-a6a5-8b52eec2b05e">

I suspect that the `$PATH` variable that you are using is clobbering the system `PATH` so it isn't finding those commands after your define it.  Maybe pick a different name for your local variable?

I am leaving tomorrow for a conference (FOSDEM) and some personal travel so may be slow to respond for about a week.  If you want to take another shot please do.  Of you you would like, I would be happy to meet and work out the last details of this one when I get back.

WarpWing commented:
> I feel like this is getting close, but I am still seeing some issues. I left a few comments in the code about specific things. Also when I run the program I'm seeing the following errors:
> <img alt="image" width="826" src="https://private-user-images.githubusercontent.com/16962482/301288413-fb10e3c5-e10c-40fd-b0bd-42a93cdbf92c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3MzQyNzMsIm5iZiI6MTcwNjczMzk3MywicGF0aCI6Ii8xNjk2MjQ4Mi8zMDEyODg0MTMtZmIxMGUzYzUtZTEwYy00MGZkLWIwYmQtNDJhOTNjZGJmOTJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMxVDIwNDYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1NDc4YjJjY2U2YTgyZTBhMThkYTJkNWY0YzI3Y2YyNjUzNmIzZjgxNzNjMGMwZDVhMjBkYTI3ZmExMWJlMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4TPGQy9Hzm7geCx3p2HCAWoQdiMfL3BNx8OwgjUQeWk">
> 
> However, both `npx` and `head` exist and run just fine.
> <img alt="image" width="826" src="https://private-user-images.githubusercontent.com/16962482/301290506-9f2bc619-dd23-400a-a6a5-8b52eec2b05e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3MzQyNzMsIm5iZiI6MTcwNjczMzk3MywicGF0aCI6Ii8xNjk2MjQ4Mi8zMDEyOTA1MDYtOWYyYmM2MTktZGQyMy00MDBhLWE2YTUtOGI1MmVlYzJiMDVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMxVDIwNDYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMDE4Y2QyMmNkZTUyN2Y4MTY1ZDBhZDI3YzU2MTc4NDBkMmZkNzA2NDUwMzBmNGVjODlmYWM4NzFjNmRmYWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gTKq_9l426_wMJZ3kosy5G1gymOfN7if397sZ4tOt8k">
> 
> I suspect that the `$PATH` variable that you are using is clobbering the system `PATH` so it isn't finding those commands after your define it. Maybe pick a different name for your local variable?
> 
> I am leaving tomorrow for a conference (FOSDEM) and some personal travel so may be slow to respond for about a week. If you want to take another shot please do. Of you you would like, I would be happy to meet and work out the last details of this one when I get back.

I'll reevaluate the system path and test that out and also add a help command. On another note, have fun at FOSDEM, I've watched some of their 2019 keynotes before on IaaS and Decentralized Internet, fascinating stuff. I'll send for a review when I'm confident this time around, I want to close this out so I can turn my attention to the conventional commit script.

WarpWing committed:
Merge branch 'development' into patch-3

WarpWing commented:
Alright, I've forced my branch back and reimplemented before those overengineered description changes. I fixed the path issue a while back, I'll send changes within the next few hours

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Removed Link and Description and node.js if statement

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Clean sweeped all env var names conflicting with system path

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Added listTargets to show all valid targets

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Reverted and modified some changes

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Removed weird flags

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Reverting to previous version

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing committed:
Update makeDocs.bash

Signed-off-by: Ty Chermsirivatana <chermsit@dickinson.edu>

WarpWing commented:
I've reverted back to a previous simple version, I think I overengineered some code and I'll put this for later.

WarpWing commented:
Closing for now. I don't want to keep this in limbo and it would be prudent for someone else to take over.
